LeanSAT/Model/Subst.lean

```diff
@@ -38,27 +46,137 @@ def PropForm.subst (p : PropForm ν₁) (f : ν₁ → PropForm ν₂) : PropFor
   | .impl φ₁ φ₂ => .impl (φ₁.subst f) (φ₂.subst f)
   | .biImpl φ₁ φ₂ => .biImpl (φ₁.subst f) (φ₂.subst f)
 
-theorem PropForm.subst_assoc (p : PropForm ν₁) (f₁ : ν₁ → PropForm ν₂) (f₂ : ν₂ → PropForm ν₃)
-  : (p.subst f₁ |>.subst f₂) = p.subst (fun v => f₁ v |>.subst f₂)
-  := by
-  induction p <;> simp [subst, *]
+section subst
 
-theorem PropForm.vars_subst [DecidableEq ν₁] [DecidableEq ν₂]
-    (p : PropForm ν₁) (f : ν₁ → PropForm ν₂)
-  : vars (p.subst f) = (vars p).biUnion (fun v1 => vars (f v1)) := by
-  induction p <;> simp [subst, Finset.biUnion_union, *]
+variable (f f₁ : ν₁ → PropForm ν₂) (f₂ : ν₂ → PropForm ν₃) (φ φ₁ φ₂ : PropForm ν₁)
 
-open PropFun in
-def PropAssignment.subst (f : ν → PropFun ν') (τ : PropAssignment ν') : PropAssignment ν :=
-  fun v => τ ⊨ f v
+@[simp] theorem subst_fls  : subst (.fls : PropForm ν₁) f = .fls                     := rfl
+@[simp] theorem subst_tr   : subst (.tr : PropForm ν₁) f = .tr                       := rfl
+@[simp] theorem subst_disj : subst (.disj φ₁ φ₂) f = .disj (subst φ₁ f) (subst φ₂ f) := rfl
+@[simp] theorem subst_conj : subst (.conj φ₁ φ₂) f = .conj (subst φ₁ f) (subst φ₂ f) := rfl
+@[simp] theorem subst_neg  : subst (.neg φ) f = .neg (subst φ f)                     := rfl
+
+theorem subst_assoc : (subst (subst φ f₁) f₂) = subst φ (fun v => subst (f₁ v) f₂) := by
+  induction φ <;> simp [subst, *]
+
+theorem vars_subst [DecidableEq ν₁] [DecidableEq ν₂]
+  : vars (φ.subst f) = (vars φ).biUnion (fun v1 => vars (f v1)) := by
+  induction φ <;> simp [subst, Finset.biUnion_union, *]
 
 @[simp]
-theorem PropForm.satisfies_subst [DecidableEq ν] (φ : PropForm ν) (f : ν → PropForm ν') {τ : PropAssignment ν'}
-  : τ ⊨ φ.subst f ↔ τ.subst (⟦f ·⟧) ⊨ φ
-  := by induction φ <;> simp [subst, PropAssignment.subst, *]; rw [PropFun.satisfies_mk]
+theorem satisfies_subst {φ : PropForm ν₁} {f} {τ : PropAssignment ν₂}
+    : τ ⊨ φ.subst f ↔ τ.subst (⟦f ·⟧) ⊨ φ := by
+  induction φ <;> simp [subst, PropAssignment.subst, *]; rw [PropFun.satisfies_mk]
+
+-- CC: Better to use equivalence or ⟦ ⟧ = ⟦ ⟧?
+theorem subst_congr {φ₁ φ₂} (hφ : equivalent φ₁ φ₂)
+    : ∀ (σ : ν₁ → PropForm ν₂), (⟦φ₁.subst σ⟧ : PropFun _) = ⟦φ₂.subst σ⟧ := by
+  intro σ
+  apply PropFun.ext
+  intro τ
+  -- CC: simp_rw doesn't work here, for some reason...
```

Comment:
I was running into the same issue, I believe it was an issue with re-instantiated typeclass arguments, which was changed in 4.7 so maybe we should update and see if it fixes that

Reply:
Comment left to address the `simp_rw` bug when we upgrade Lean version.